design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,147 @@
+wants to use Kubernetes manifests for a declarative intent for a Ceph cluster, and the Ceph
+Orchestrator Mgr Module, which treats OSD creation as imperative and allows the on-disk content of
+OSDs to be the declarative state.
+
+### Adding OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section for individual `nodes` to have disks added via a Drive Groups specification. The spec is
+proposed below with behavior described to follow.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage:
+    nodes: # preexisting
+      - name: < node name > # preexisting
+        driveGroups: # proposed new
+          - removeFromSpecAfterCompmletion: < default false > # proposed new
+            spec: < JSON string to pass to ceph-volume for OSD creation >
+```
+
+Each individual `driveGroups` item's `spec` will be a simple string containing a JSON blob that gets
+passed along without modification or inspection by Rook to `ceph-volume` when running the OSD
+provisioning job on the corresponding node. The `ceph-volume` command to create OSDs via the Drive
+Groups spec is given priority and executed first before any other Rook-based OSD creation methods
+are used.
+
+Users using `CephCluster` via Kubernetes Manifest will now be able to supply their own Drive Groups
+spec to Rook if they so desire to allow themselves greater flexibility in OSD management. Rook will
+treat this user spec as a declarative intent for the cluster so long as the
+`removeFromSpecAfterCompletion` is unspecified or set to `false`. Setting this value to `true` will
+change Rook's behavior to treat the spec imperatively, intended for use by the Orchestrator Mgr
+Module. Upon completion of the `ceph-volume` command with the Drive Group spec, Rook will remove the
+Drive Group item from the `driveGroups` spec for the corresponding node in the `CephCluster`
+resource so that Rook will not rerun the creation again, which would break the Ceph Orchestrator Mgr
+Module's imperative intention. The Mgr Module should obviously set this value to `true` for all
+operations.
+
+**TODO DISCUSS THIS:**
+The Drive Group item should be removed regardless of whether the `ceph-volume` command result was
+successful or unsuccessful. Rook should report the success/fail result back to the Rook
+Orchestrator Mgr Module via a `CephCluster` status mechanism so that the Orchestrator can take
+appropriate actions to notify the user of failure. The status should be cleared by the Rook
+Orchestrator Mgr Module when the user dismisses the status message, and Rook should never
+clear statuses on its own, only appending new statuses to a list.
+
+```yaml
+# first draft suggestions for status mechanism
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+status: # preexisting
+  driveGroupOSDCreations: # proposed new
+    - node: <node name> # proposed new
+      driveGroupSpec: < JSON string to pass to ceph-volume for OSD creation >
```

Comment:
With the proposal to keep the driveGroup in the CR, we just need a unique identifier for the driveGroup here.

Reply:
Why is there a need for a unique identifier? This is a list and not a dict. I would assume that the JSON string is as much a unique identifier as anything, no? The user could theoretically use the same string twice, but since `ceph-volume` is idempotent (or should be), it won't break provisioning as long as we make sure to only manipulate one list item at a time.

Reply:
It should be possible to correlate a drive group between the `storage` spec and the `status` spec by the spec string JSON (or YAML?) and the node name. Internally, an ID could be a hash of struct including these 2 components I suppose.

Reply:
I like the hash as an identifier. Including a whole json/yaml blob just doesn't feel right for the status section.